lib/pacemaker/pcmk_resource.c

```diff
@@ -103,6 +104,68 @@ best_op(const pcmk_resource_t *rsc, const pcmk_node_t *node)
     return best;
 }
 
+/*!
+ * \internal
+ * \brief Remove a resource
+ *
+ * \param[in,out] cib       An open connection to the CIB
+ * \param[in]     cib_opts  Options to use in the CIB operation call
+ * \param[in]     rsc_id    Resource to remove
+ * \param[in]     rsc_type  Type of the resource ("primitive", "group", etc.)
+ *
+ * \return Standard Pacemaker return code
+ */
+int
+pcmk__resource_delete(cib_t *cib, uint32_t cib_opts, const char *rsc_id,
+                      const char *rsc_type)
+{
+    int rc = pcmk_rc_ok;
+    xmlNode *msg_data = NULL;
+
+    if (cib == NULL) {
+        return ENOTCONN;
+    }
+
+    if (rsc_id == NULL || rsc_type == NULL) {
+        return EINVAL;
+    }
+
+    msg_data = create_xml_node(NULL, rsc_type);
+    crm_xml_add(msg_data, XML_ATTR_ID, rsc_id);
+
+    rc = cib->cmds->remove(cib, XML_CIB_TAG_RESOURCES, msg_data, cib_opts);
+    rc = pcmk_legacy2rc(rc);
+
+    free_xml(msg_data);
+    return rc;
+}
+
+int
+pcmk_resource_delete(const char *rsc_id, const char *rsc_type)
```

Comment:
A couple things to note on this:

1. I can't find any instance of a public libpacemaker function handling an xmlNodePtr argument without also using pcmk__xml_output_new (and therefore, a pcmk__output_t) at the same time.  Deleting doesn't do any output, so this is a little weird, but whatever.  I just think it's interesting this hasn't come up before.
2. Several of the public pcmk_fence_* functions do not take an xmlNodePtr - seems like we need to clean that up.
3. I don't think we even handle the possibility of failure right now, unless somewhere along the lines, something outputs an error.  For instance, look at pcmk_resource_digests.  It calls pcmk__resource_digests, which can return various error codes.  Those error codes will be returned by the public function, but nowhere will they end up in the XML node because pcmk__xml_output_finish doesn't take the error code as a parameter.  It previously did, but it also used that to decide whether to do any output at all.  Maybe it makes sense to take the error code and add it (with out->error?).

Reply:
Yep, we should definitely create an XML output object for all public APIs. Even if there is no output, the XML will have the exit status code and message. It lets users have consistent handling for all calls. The XML definitely has a `<status code="N" message="BLAH">` element; I'm not sure whether that's being set somewhere else or we're always returning success for digests. At least crm_attribute can return an error code there.

Delete probably *should* have success/failure output. Maybe for 3.0.0. 

The fence APIs aren't really public, they're guarded by an undefined constant. `stonith_t` will eventually be replaced by`pcmk_ipc_api_t`, so I don't want to expose anything new using it. When we get to that point, we can either make the fence APIs take a `pcmk_ipc_api_t`, or just create it within each function like we do with other connections.

Reply:
> Yep, we should definitely create an XML output object for all public APIs. Even if there is no output, the XML will have the exit status code and message. It lets users have consistent handling for all calls. The XML definitely has a `<status code="N" message="BLAH">` element; I'm not sure whether that's being set somewhere else or we're always returning success for digests. At least crm_attribute can return an error code there.

Right, I think this is the difference between using the public and private functions.

In a tool like crm_attribute, we are using the private functions (or, we would be if we're using any at all - I haven't checked).  We create a pcmk__output_t, call some function, and get a return value.  At the end, we call `out->finish()` with that return value.  It's `xml_finish` that constructs the wrapper around the rest of the XML with the return code and message.

If we're using the public functions, we do not have a pcmk__output_t to begin with.  The public function creates one, calls a function, and gets a return value.  We call `pcmk__xml_output_finish` which doesn't take a return code argument, which then calls `out->finish()` and only ever passes 0 for the return code.  So, public function XML will always be wrapped up with success.  The return code gets passed back to the caller of the public function which can still exit with error, but the output has already happened.

Long story short, I think `pcmk__xml_output_finish` needs to take a return code and pass that through to `out->finish()`.

Reply:
See #3270.